### PR TITLE
OrderedIntervalList

### LINKED ISSUE #41: Solidity library OrderedLinkedList
- Use `library` with struct and all `internal` methods
- Allow to add ranges (check order and collisions)
- Allow to remove ranges (check existence)

### PR DESCRIPTION
**Pull request by patsak**
resolve #41

### CODE REVIEW

File: software/smartContract/contracts/OrderedIntervalList.sol [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        return self.intervals.length;
+    }
+
+    /**
+    @notice Get interval by list index
+    @param _index interval index in list
+    @return interval tuple
+     */
+    function get(Data storage self, uint _index) internal view returns(uint64 begin, uint64 end) {
+        require(self.intervals.length > 0 && _index < self.intervals.length, "check list is non empty and index have proper bounds");
```
Comment (Eenae) [inline software/smartContract/contracts/OrderedIntervalList.sol]: Here and below: message parameter of a `require` function is a string message containing details about the error.

File: software/smartContract/contracts/OrderedIntervalList.sol [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        Interval memory interval = self.intervals[_index];
```
Comment (Eenae) [inline software/smartContract/contracts/OrderedIntervalList.sol]: s/memory/storage/ ?

File: software/smartContract/contracts/OrderedIntervalList.sol [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        nextRawIndex = _index + 1;        
+        while (nextRawIndex < self.intervals.length && next == 0) {
+            Interval memory nextEl = self.intervals[nextRawIndex];
+            if (nextEl.end != 0) {
+                next = nextEl.end;
+            }
+            nextRawIndex = nextRawIndex + 1;                        
+        }       
+
+        require((prev <= _begin || prev == 0) && (next == 0 || next >= _end), "valid interval bounds");
```
Comment (Eenae) [inline software/smartContract/contracts/OrderedIntervalList.sol]: Why are't we iterating `prev` backwards as in case of iteration over `next`?

Comment (Eenae) [reply software/smartContract/contracts/OrderedIntervalList.sol]: test case: `(0, 10), (0, 0), (0, 0), (20, 30)`, inserting `(5, 15)` at index 2

File: software/smartContract/contracts/OrderedIntervalList.sol (hunk 1 above)
Comment (k06a) [reply software/smartContract/contracts/OrderedIntervalList.sol]: Remove first part of this condition, second part coverts this case.

File: software/smartContract/contracts/OrderedIntervalList.sol (hunk 2 above)
Comment (k06a) [reply software/smartContract/contracts/OrderedIntervalList.sol]: Maybe better to return whole struct, compile will split it automagically.

**Comment by k06a** [comment]
Arguments should not have underscore prefix anymore, see OpenZeppelin 2.0 notation.

File: software/smartContract/contracts/OrderedIntervalList.sol [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+            self.intervals[_index] = Interval({begin: _begin, end: _end});
+        }
+        
+    } 
+
+    /**
+        @notice Remove interval by index
+        @param _index interval index in list        
+     */
+    function remove(Data storage self, uint64 _index) internal {
```
[the remark this answers is not in the document]
Comment (patsak) [reply software/smartContract/contracts/OrderedIntervalList.sol]: So, is it  need an additional method like remove(Data self, uint64 begin, uint end)? Or replace existed "remove" method?
Does "remove by range" method have to remove ranges that have any intersection with this range or only full intersection?

**Pull request merged**